include/zephyr/net/coap_service.h

```diff
@@ -0,0 +1,255 @@
+
+/**
+ * @brief Iterate over all static resources associated with @p _service .
+ *
+ * @note This macro is suitable for a @p _service defined with @ref COAP_SERVICE_DEFINE.
+ *
+ * @param _service Pointer to COAP service
+ * @param _it Name of iterator (of type @ref coap_resource)
+ */
+#define COAP_SERVICE_FOREACH_RESOURCE(_service, _it)						\
+	for (struct coap_resource *_it = (_service)->res_begin; ({				\
+		__ASSERT(_it <= (_service)->res_end, "unexpected list end location");		\
+		_it < (_service)->res_end;							\
+	}); _it++)
+
+/**
+ * @brief Start the provided @p service .
+ *
+ * @note This macro is suitable for a @p service defined with @ref COAP_SERVICE_DEFINE.
+ *
+ * @param service Pointer to CoAP service
+ * @retval 0 in case of success.
+ * @retval -EALREADY in case of an already running service.
+ * @retval -ENOMEM in case the server has no available context.
+ */
+int coap_service_start(const struct coap_service *service);
+
+/**
+ * @brief Stop the provided @p service .
+ *
+ * @note This macro is suitable for a @p service defined with @ref COAP_SERVICE_DEFINE.
+ *
+ * @param service Pointer to CoAP service
+ * @retval 0 in case of success.
+ * @retval -EALREADY in case the service isn't running.
+ */
+int coap_service_stop(const struct coap_service *service);
+
+/**
+ * @brief Send a CoAP message from the provided @p service .
+ *
+ * @note This macro is suitable for a @p service defined with @ref COAP_SERVICE_DEFINE.
+ *
+ * @param service Pointer to CoAP service
+ * @param cpkt CoAP Packet to send
+ * @param addr Peer address
+ * @param addr_len Peer address length
+ * @return 0 in case of success or negative in case of error.
+ */
+int coap_service_send(const struct coap_service *service, const struct coap_packet *cpkt,
+		      const struct sockaddr *addr, socklen_t addr_len);
+
+/**
+ * @brief Send a CoAP message from the provided @p resource .
+ *
+ * @param resource Pointer to CoAP resource
+ * @param cpkt CoAP Packet to send
+ * @param addr Peer address
+ * @param addr_len Peer address length
+ * @return 0 in case of success or negative in case of error.
+ */
+int coap_resource_send(const struct coap_resource *resource, const struct coap_packet *cpkt,
+		       const struct sockaddr *addr, socklen_t addr_len);
+
+/**
+ * @brief Parse a CoAP observe request for the provided @p resource .
+ *
+ * @param resource Pointer to CoAP resource
+ * @param request CoAP request to parse
+ * @param addr Peer address
+ * @return 0 in case of success or negative in case of error.
+ */
+int coap_resource_parse_observe(struct coap_resource *resource, const struct coap_packet *request,
+				const struct sockaddr *addr);
+
+/**
+ * @brief Lookup an observer by address and remove it from the @p resource .
+ *
+ * @param resource Pointer to CoAP resource
+ * @param addr Peer address
+ * @retval 1 in case an observer was removed.
+ * @retval 0 in case no observer was removed.
+ * @return negative value in case of an error.
```

Comment:
Mixing `retval` with `return` does not render well in API docs, we end up with two "Return" sections:
https://builds.zephyrproject.io/zephyr/pr/64265/docs/doxygen/html/group__coap__service.html#ga8d9ab0bf6b1ea15408f1c80c45aae16b

Plus, I wonder about the return values, wouldn't it make sense to return `0` in case of success (observer removed) and treat missing observer as an error condition `-ENOENT`?

Reply:
Updated the documentation, and applied te suggestion.